### ISSUE TITLE
Token request fails with DeviceManagementScreenlockRequired

### ISSUE THREAD

**Issue by mrjackv** [microg/GmsCore#896]
As the title states, when GmsAuthManagerSvc tries to make a request for a new token it fails.   

This is the stacktrace:
```
W GmsAuthManagerSvc: 
W GmsAuthManagerSvc: java.io.IOException: Error=DeviceManagementScreenlockRequired
W GmsAuthManagerSvc:     at org.microg.gms.common.HttpFormClient.request(HttpFormClient.java:90)
W GmsAuthManagerSvc:     at org.microg.gms.auth.AuthRequest.getResponse(AuthRequest.java:194)
W GmsAuthManagerSvc:     at org.microg.gms.auth.AuthManager.requestAuth(AuthManager.java:226)
W GmsAuthManagerSvc:     at org.microg.gms.auth.AuthManagerServiceImpl.getToken(AuthManagerServiceImpl.java:95)
W GmsAuthManagerSvc:     at org.microg.gms.auth.AuthManagerServiceImpl.getTokenWithAccount(AuthManagerServiceImpl.java:160)
W GmsAuthManagerSvc:     at com.google.android.auth.IAuthManagerService$Stub.onTransact(IAuthManagerService.java:134)
W GmsAuthManagerSvc:     at org.microg.gms.auth.AuthManagerServiceImpl.onTransact(AuthManagerServiceImpl.java:175)
W GmsAuthManagerSvc:     at android.os.Binder.execTransact(Binder.java:731)
```
My first hunch was to check that I had my screen lock PIN enabled, and it was. Other than that I tried finding other issues/information but it seems that I'm the only one with this problem.

I am currently using LineageOS 16 with microg, and have manually updated GmsCore to version 0.2.8.17785-4 (06c8b76)

**Comment by benwaffle** [comment]
I am getting this same error with GMail and FairEmail

**Comment by Coo-ops** [comment]
Bump please, its a 4 year old bug that renders custom ROMs useless for Google Workspace businesses.

**Comment by alexwh** [comment]
> Bump please, its a 4 year old bug that renders custom ROMs useless for Google Workspace businesses.

See https://old.reddit.com/r/Vanced/comments/kzqyww/question_cannot_sign_in_as_of_a_few_days_ago_and/hmzqd41/?context=3

**Comment by Coo-ops** [comment]
> > Bump please, its a 4 year old bug that renders custom ROMs useless for Google Workspace businesses.
> 
> See https://old.reddit.com/r/Vanced/comments/kzqyww/question_cannot_sign_in_as_of_a_few_days_ago_and/hmzqd41/?context=3

Not a useful comment, I am not my company's Google admin.

This needs to be fixed, its a dumb issue.

**Comment by aximut** [comment]
Since nobody answered my questions above, I spent some time to see what is missing. Same as for #1726.
It is EXACTLY 2 requests.

The original GMS app also gets these mentioned `DeviceManagementScreenlockRequired` errors on requests to `https://android.googleapis.com/auth`, but in contrast to the microG implementation, it then does the following two requests and then the auth requests are repeated and everything regarding auth gets no more 403 errors, but works fine:

- **First request**:
A gRPC POST request to
`https://cryptauthenrollment.googleapis.com/google.cryptauth.enrollment.v1.Enrollment/SyncKeys`
(the black parts are all signatures and tokens individual to the device/registration)
**I am assuming that one of these uint32 is telling whether the lockscreen is enabled or not**
![SyncKeys](https://github.com/microg/GmsCore/assets/27779900/bf6a26ff-2fe7-4535-acd5-68a86a6eff20)
- **Second request**:
 A gRPC POST request to
 `https://cryptauthenrollment.googleapis.com/google.cryptauth.enrollment.v1.Enrollment/EnrollKeys`
 (All these `b'\...` messages are very long hex byte strings)
![EnrollKeys](https://github.com/microg/GmsCore/assets/27779900/20f820c7-a75c-475f-b19b-63aae0bac96a)


These requests repeat themselves more often afterwards with different data (always first `SyncKeys`, then `EnrollKeys`), but the two requests mentioned here are sufficient for the exception of this issue to disappear and for Goolge Workspace and all related apps (Chat, Docs, ...) to be fully working.
By the sheer amount of auth tokens and signatures that need to be generated and valid for these requests to succeed, I understand why this is not implemented yet. Also, the error messages you get from these gRPC endpoints are not helpful in debugging either. The source to look into how these signatures are generated is of course the official implementation.

**What is also very interesting is that once these requests have been done for one Workspace account once on the device, they DO NOT need to be repeated. The account can be removed and a totally different Workspace account can be added. The requests above are not needed anymore. They are really only required once per device to register. That also makes debugging hard because you need to totally reset the device to provoke the error again**.

**Comment by aximut** [comment]
Since nobody answered my questions above, I spent some time to see what is missing. Same as for #1726.
It is EXACTLY 2 requests.

The original GMS app also gets these mentioned `DeviceManagementScreenlockRequired` errors on requests to `https://android.googleapis.com/auth`, but in contrast to the microG implementation, it then does the following two requests and then the auth requests are repeated and everything regarding auth gets no more 403 errors, but works fine:

- **First request**:
A gRPC POST request to
`https://cryptauthenrollment.googleapis.com/google.cryptauth.enrollment.v1.Enrollment/SyncKeys`
(the black parts are all signatures and tokens individual to the device/registration)
**I am assuming that one of these uint32 is telling whether the lockscreen is enabled or not**
![SyncKeys](https://github.com/microg/GmsCore/assets/27779900/29549514-dc3c-463c-bce0-28a0e5f2b3ad)
- **Second request**:
 A gRPC POST request to
 `https://cryptauthenrollment.googleapis.com/google.cryptauth.enrollment.v1.Enrollment/EnrollKeys`
 (All these `b'\...` messages are very long hex byte strings)
![EnrollKeys](https://github.com/microg/GmsCore/assets/27779900/20f820c7-a75c-475f-b19b-63aae0bac96a)


These requests repeat themselves more often afterwards with different data (always first `SyncKeys`, then `EnrollKeys`), but the two requests mentioned here are sufficient for the exception of this issue to disappear and for Goolge Workspace and all related apps (Chat, Docs, ...) to be fully working.
By the sheer amount of auth tokens and signatures that need to be generated and valid for these requests to succeed, I understand why this is not implemented yet. Also, the error messages you get from these gRPC endpoints are not helpful in debugging either. The source to look into how these signatures are generated is of course the official implementation.

**What is also very interesting is that once these requests have been done for one Workspace account once on the device, they DO NOT need to be repeated. The account can be removed and a totally different Workspace account can be added. The requests above are not needed anymore. They are really only required once per device to register. That also makes debugging hard because you need to totally reset the device to provoke the error again**.

**Comment by fynngodau** [comment]
Thanks a lot for having a look @aximut. I'm currently having a very intense look at resolving this issue; here's some information about the `SyncKeys` query you shared:

* grpc: to decode the raw query data manually, you can use https://gist.github.com/fynngodau/bba7b2d11a17d01f8439979f3cfe4fb9
* Relevant protobuf: https://github.com/chromium/chromium/blob/main/chromeos/ash/services/device_sync/proto/cryptauth_enrollment.proto#L31 leads to very reasonable results
* For `client_app_metadata` field, the relevant protobuf should be https://github.com/chromium/chromium/blob/main/chromeos/ash/services/device_sync/proto/cryptauth_client_app_metadata.proto#L16
* specifically in that protobuf (in your screenshot: 6.15), [`using_secure_screenlock`](https://github.com/chromium/chromium/blob/8eb02fd1c2c64b404a0d593a88bdd71239dca764/chromeos/ash/services/device_sync/proto/cryptauth_client_app_metadata.proto#L69) looks very much like the field we need according to my tests

The above proto files are a bit hard to use because field 6 of `SyncKeysRequest` is `bytes`, not `ClientAppMetadata`. With [this combined proto file](https://gist.github.com/fynngodau/04864facebf33001ad172f10260bf0f4), decoding the sync request works beautifully:

```
$ protoc --decode=cryptauthv2.SyncKeysRequest cryptauth_combined.proto < cryptauth-synckeys-with-screenlock.dec
```

**Comment by aximut** [comment]
Thanks for sharing and putting everything in a useful file! That is very interesting that the protobufs are part of the chromium repo. I did not find any references when I searched it back in the day.

In terms of complexity it is as I suspected, lot's of crypto going on. Especially parts like [`KeyStorageLevel`](https://gist.github.com/fynngodau/04864facebf33001ad172f10260bf0f4#file-cryptauth_combined-proto-L113) and [`KeyAttestation`](https://gist.github.com/fynngodau/04864facebf33001ad172f10260bf0f4#file-cryptauth_combined-proto-L195) probably require quite some dive into the Android Keystore and TEE storage to provide proper attestation to the backend, depending on the enforcement rules on the Google servers.

**Comment by fynngodau** [comment]
@aximut It's not really as complex, many of the intimidating fields can be removed and the `EnrollKeys` step is optional. If you know the Android ID that microG uses for you (you can sometimes see it in the queries it makes and logs to logcat) and insert it into a decoded query as above in `message.txt` (of course, `using_secure_screenlock` should be set to the correct value), you can use [`pack.py`](https://gist.github.com/fynngodau/eebf750dac2be8440b231767a4f1da67) with the following to tell the Google server that your microG device has screenlock enabled:

```
protoc --encode=cryptauthv2.SyncKeysRequest cryptauth_combined.proto < message.txt > message.prot
python pack.py message.prot
```

With that, you have a file `message.prot.pac` that you can send as a request body for the same request as the one you intercepted. For example, you can upload this file in mitmproxy and click Replay. Afterwards, `DeviceManagementScreenlockRequired` error will no longer appear.

I'll be working on a microG implementation next.

**Comment by ale5000-git** [comment]
> What is also very interesting is that once these requests have been done for one Workspace account once on the device, they DO NOT need to be repeated. The account can be removed and a totally different Workspace account can be added. The requests above are not needed anymore. They are really only required once per device to register. That also makes debugging hard because you need to totally reset the device to provoke the error again

Not necessarily.
Maybe only removing the account isn't enough but using my script in this way `minutil.sh --remove-all-accounts` will remove all account related things (need root) and may work.

The script is here: https://github.com/micro5k/microg-unofficial-installer/blob/main/zip-content/files/bin/minutil.sh

**Comment by fynngodau** [comment]
You can simply set `using_secure_screenlock` in above queries to false again for the Android ID in question, and the problem will reoccur (and, seemingly, the account deauthenticated such that it needs to be readded).

On Google Play services devices, the query does happen after each account setup.

**Comment by aximut** [comment]
> It's not really as complex, many of the intimidating fields can be removed and the `EnrollKeys` step is optional for us.

That is good news. Didn't expect Google to be that lax on the requirements.

> With that, you have a file `message.prot.pac` that you can send as a request body for the same request as the one you intercepted.

Ok, that means next to the Android ID we only need to get the other headers (`authorization`, `x-goog-api-key`, ...) right for it to succeed.

> On Google Play services devices, the query does happen after each account setup.

Yes, I saw it but it stays already authenticated even before the call hits first time. The calls are going on in quite an async manner in the official implementation.

**Comment by fynngodau** [comment]
@aximut Draft PR for this issue: https://github.com/microg/GmsCore/pull/2296

**Comment by aximut** [comment]
Thanks a lot for making this happen @fynngodau !
Will test this as soon as possible.

**Comment by aximut** [comment]
Just built your branch on [GitHub actions](https://github.com/aximut/android_packages_apps_GmsCore/actions/runs/8561696321). Using the debug build. Getting the following logcat when clicking on the account settings screen (only relevant parts filtered).
Basically a short spinner appears when clicking and then it returns back to the previous screen. No account settings displayed, and so far could not get the DeviceManagementScreenlockRequired to disappear.

```
D AccountSettings: Invoked with com.google.android.gms.accountsettings.SECURITY_SETTINGS and extras Bundle[{account=Supplier{VAL_PARCELABLE@24+228}}]
I ActivityTaskManager: Displayed com.google.android.gms/org.microg.gms.accountsettings.ui.MainActivity: +135ms
D GmsHttpFormClient: -- Request --
D GmsHttpFormClient: androidId=...
...
W AccountSettingsWebView: Failed to get weblogin auth.
W AccountSettingsWebView: java.io.IOException: Error=DeviceManagementScreenlockRequired
W AccountSettingsWebView: 	at org.microg.gms.common.HttpFormClient.request(HttpFormClient.java:96)
W AccountSettingsWebView: 	at org.microg.gms.auth.AuthRequest.getResponse(AuthRequest.java:258)
W AccountSettingsWebView: 	at org.microg.gms.auth.AuthManager.requestAuth(AuthManager.java:293)
W AccountSettingsWebView: 	at org.microg.gms.accountsettings.ui.WebViewHelper.openWebWithAccount(WebViewHelper.kt:109)
W AccountSettingsWebView: 	at org.microg.gms.accountsettings.ui.WebViewHelper.access$openWebWithAccount(WebViewHelper.kt:32)
...
D CoreBackPreview: Window{f7e791 u0 Splash Screen com.google.android.gms EXITING}: Setting back callback null
```

**Comment by simzen85** [comment]
@aximut @fynngodau may I ask when this fix is merged ? Tks

Seems to affect the ability to sync with Business Google email

**Comment by fynngodau** [comment]
@simzen85 The PR can be merged only after it is working. See the PR for details.

**Comment by fynngodau** [comment]
The feature to support Workspace accounts with basic integrity is included in v0.3.4. I suggest to close this issue.

**Comment by ale5000-git** [comment]
I'm closing it, if someone has still problems with the latest version please write here.

**Comment by Caquin1950** [comment]
Hey, so I've been unable to use my Google Workspace account using the Huawei version of MicroG for the past year. I found this thread and updated to the latest v0.3.6.244735-hw, but the problem persists. Trying to load anything with this accounts results in infinite loading or "check your connection" messages, while other google accounts work fine.

I'm tech savy enough, but I'm not familiar with Android development. That said, I'm willing to dig in as much as needed in order to cooperate fixing this issue, I would greatly appreciate it since this issue has been making me use some tiresome workarounds, it would be great if it could be fixed! Thanks :)

**Comment by ale5000-git** [comment]
Recently the PR #2553 was merged, maybe it contains something related to your problem.

**Comment by Caquin1950** [comment]
Will do in a few minutes! Took me a bit to figure out how to output the proper logs, but I'll open a new issue. Thank you very much!
